addons/mrp/models/mrp_bom.py

```diff
@@ -192,34 +192,44 @@ def _bom_find(self, product_tmpl=None, product=None, picking_type=None, company_
         """Optimized variant of _bom_find to work with recordset"""
-        products = products.filtered(lambda product: product.type != 'service')
+        non_service_products = products.filtered(lambda p: p.type != 'service')
         if not products:
-            return {}
-        product_templates = products.mapped('product_tmpl_id')
-        domain = ['|', ('product_id', 'in', products.ids), '&', ('product_id', '=', False), ('product_tmpl_id', 'in', product_templates.ids)]
-        if self.env.context.get('company_id'):
-            domain = domain + ['|', ('company_id', '=', False), ('company_id', '=', self.env.context.get('company_id'))]
+            return self.env['mrp.bom']
+        if not non_service_products:
+            return [False for product in products]
+        product_templates = non_service_products.mapped('product_tmpl_id')
+        domain = ['|', ('product_id', 'in', non_service_products.ids), '&', ('product_id', '=', False), ('product_tmpl_id', 'in', product_templates.ids)]
+        if picking_type:
+            domain += ['|', ('picking_type_id', '=', picking_type.id), ('picking_type_id', '=', False)]
+        if company_id or self.env.context.get('company_id'):
+            domain += ['|', ('company_id', '=', False), ('company_id', '=', company_id or self.env.context.get('company_id'))]
         if bom_type:
             domain += [('type', '=', bom_type)]
 
         boms = self.search(domain, order='sequence, product_id')
         template2bom = {}
         variant2bom = {}
+        # Cannot use bom.sequence as unset sequence values are set to 0.
+        # Use counter to maintain the order of self.search above.
+        psql_sequence = 0
         for bom in boms:
             # Use "setdefault" to take only first bom if we have few ones for
             # the same product
             if bom.product_id:
-                variant2bom.setdefault(bom.product_id, bom)
+                variant2bom.setdefault(bom.product_id, {'bom': bom, 'sequence': psql_sequence})
             else:
-                template2bom.setdefault(bom.product_tmpl_id, bom)
-
-        result = {}
+                template2bom.setdefault(bom.product_tmpl_id, {'bom': bom, 'sequence': psql_sequence})
+            psql_sequence = psql_sequence + 1
+        result = []
         for p in products:
-            bom = variant2bom.get(p) or template2bom.get(p.product_tmpl_id)
-            if bom:
-                result[p] = bom
+            if p.type == 'service':
+                result.append(False)
+            elif template2bom.get(p.product_tmpl_id, {'sequence': float('inf')})['sequence'] < variant2bom.get(p, {'sequence':float('inf')})['sequence']:
+                result.append(template2bom.get(p.product_tmpl_id, {}).get('bom'))
+            else:
+                result.append(variant2bom.get(p, {}).get('bom') or template2bom.get(p.product_tmpl_id, {}).get('bom'))
```

Comment:
That's because `_get_product2bom` is basically a batch version of `_bom_find`. `_bom_find` is supposed to return 1 bom for a given product, with the `sequence` param being the first discriminant.

If we have the following data

| bom_id | sequence | product_tmpl_id | product_id |
|:-------:|:----------:|:-----------------:|:-----------:|
| 1|0|1| |
| 2|1|1|1|

Since `mrp.bom(1)` has a lower sequence, `_bom_find` will return `mrp.bom(1)` for product.product(1). Currently, `_get_product2bom` will return `mrp.bom(2)` because it prioritizes boms with `product_id != False` while not accounting for `sequence` values.

This did not cause any issue when merging `_get_product2bom` because `product._computate_quantities_dict` didn't use that. But there are some unit tests for `mrp.bom.explode` that failed because of that.

Reply:
I see, thanks for the clarification